### PR TITLE
fix(workspace): Update backlinks after engine updates

### PR DESCRIPTION
**Pull request by tma66**
**Background**
Currently we do not update backlinks when a note is created/updated/deleted. This pr fills in that gap by computing the diff of links between the old and new note. This will be implemented for both the engine v3 and v2.

**Changes**
1. For a newly created note, scan through its links and create backlink if applicable
2. For a deleted note,  scan through its links and delete backlink if applicable
3. For updated note, calculate diff and add/remove backlinks
4. Created new `BacklinkUtils` to handle basic backlink methods. `LinkUtils` is intertwined with many other services so I wanted to keep it separate

### CODE REVIEW

File: packages/common-all/src/engine/EngineV3Base.ts [hunk 1]
```diff
@@ -43,7 +46,8 @@ export abstract class EngineV3Base implements ReducedDEngine {
 
   constructor(
     protected noteStore: INoteStore<string>,
-    protected logger: DLogger
+    protected logger: DLogger,
+    public vaults: DVault[]
```
Comment (jonathanyeung) [inline packages/common-all/src/engine/EngineV3Base.ts:50]: I'm thinking it might be better to make vaults `protected` instead of `public`.  Not sure if using the engine is the right place to get `vaults`.  Is this possible, or does it lead to a snowball effect of changes?

File: packages/common-all/src/engine/EngineV3Base.ts [hunk 2]
```diff
@@ -382,4 +392,88 @@ export abstract class EngineV3Base implements ReducedDEngine {
...
+      }
+    }
+    return [];
+  }
+
+  /**
+   * Remove backlink associated with given link that references another note (denoted by presence of link.to field)
+   * from that referenced note
+   *
+   * @param link Link potentionally referencing another note
```
Comment (jonathanyeung) [inline packages/common-all/src/engine/EngineV3Base.ts:442]: nit: sp potentially

File: packages/common-all/src/engine/EngineV3Base.ts [hunk 3]
```diff
@@ -382,4 +392,88 @@ export abstract class EngineV3Base implements ReducedDEngine {
...
+   */
+  protected async removeBacklink(link: DLink): Promise<NoteChangeEntry[]> {
```
Comment (jonathanyeung) [inline packages/common-all/src/engine/EngineV3Base.ts:444]: aside - should we maybe undeprecate `DLink`? It's been marked as deprecate for as long as I can remember and its suggested replacement `DNoteLink` is practically the same and that interface needs some cleanup too.  @kevinslin

File: packages/common-all/src/engine/EngineV3Base.ts [hunk 4]
```diff
@@ -382,4 +392,88 @@ export abstract class EngineV3Base implements ReducedDEngine {
...
+    const maybeBacklink = BacklinkUtils.createFromDLink(link);
+    if (maybeBacklink) {
+      const maybeVault = link.to?.vaultName
+        ? VaultUtils.getVaultByName({
+            vname: link.to?.vaultName,
+            vaults: this.vaults,
+          })
+        : undefined;
+      const notes = await this.noteStore.findMetaData({
+        fname: link.to!.fname!,
```
Comment (jonathanyeung) [inline packages/common-all/src/engine/EngineV3Base.ts:454]: instead of `!` - can we do that same thing again as in the other review? If those aren't defined, do an early return of [] to save some backend calls.

File: packages/common-all/src/engine/EngineV3Base.ts [hunk 5]
```diff
@@ -382,4 +392,88 @@ export abstract class EngineV3Base implements ReducedDEngine {
...
+        vault: maybeVault,
+      });
+      if (notes.data) {
+        return Promise.all(
+          notes.data.map(async (note) => {
+            const prevNote = _.merge(_.cloneDeep(note), {
+              body: "",
```
Comment (jonathanyeung) [inline packages/common-all/src/engine/EngineV3Base.ts:461]: hm - should `NoteChangeEntry` be changed to use propsMeta instead? not sure - something to consider later.

File: packages/common-all/src/engine/EngineV3Base.ts [hunk 6]
```diff
@@ -382,4 +392,88 @@ export abstract class EngineV3Base implements ReducedDEngine {
...
+    const maybeBacklink = BacklinkUtils.createFromDLink(link);
+    if (maybeBacklink) {
+      const maybeVault = link.to?.vaultName
+        ? VaultUtils.getVaultByName({
+            vname: link.to?.vaultName,
+            vaults: this.vaults,
+          })
+        : undefined;
+      const notes = await this.noteStore.findMetaData({
+        fname: link.to!.fname!,
```
Comment (jonathanyeung) [inline packages/common-all/src/engine/EngineV3Base.ts:412]: same comment as below.

File: packages/unified/src/remark/utils.ts [hunk 7]
```diff
@@ -365,6 +366,45 @@ const getLinkCandidates = async ({
 };
 
 export class LinkUtils {
+  static isEquivalent(linkA: DLink, linkB: DLink): boolean {
```
Comment (jonathanyeung) [inline packages/unified/src/remark/utils.ts:369]: These methods seem more appropriate somewhere in `common-all`.

Same can be said about a lot of methods in this class... we should break down this file into methods that need to be in `unified` package and those that don't. Maybe in the next iteration of refactoring.

File: packages/common-all/src/BacklinkUtils.ts [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+   *  @param note note that the link is pointing to. (mutated)
+   *  @param link backlink to add. */
+  static addBacklink({
+    note,
+    backlink,
+  }: {
+    note: NotePropsMeta;
+    backlink: Omit<DLink, "type"> & { type: "backlink" };
+  }): void {
+    note.links.push(backlink);
```
Comment (jonathanyeung) [inline packages/common-all/src/BacklinkUtils.ts:37]: any chance we have duplicates?

File: packages/plugin-core/src/test/suite-integ/CopyNoteLink.test.ts [hunk 9]
```diff
@@ -366,7 +366,7 @@ suite("CopyNoteLink", function () {
             engine,
           });
 
-        const editor = await openNote(noteWithTarget);
+        let editor = await openNote(noteWithTarget);
```
Comment (jonathanyeung) [inline packages/plugin-core/src/test/suite-integ/CopyNoteLink.test.ts:369]: was this change intentional? maybe I'm reading it wrong, but the second `openNote()` call seems redundant?

**Review by jonathanyeung — CHANGES_REQUESTED** [review]
Wondering if this will maybe fix this issue? It's our most recent CROP selection: https://github.com/dendronhq/dendron/issues/1207

If not, then any changes I do on it I'll build off this review.

File: packages/common-all/src/BacklinkUtils.ts [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+export class BacklinkUtils {
+  /**
+   * Create backlink out of link if it references another note (denoted by presence of link.to field)
+   *
+   * @param link Original link to create backlink out of
+   * @returns backlink or none if not applicable
+   */
+  static createFromDLink(
+    link: DLink
+  ): (Omit<DLink, "type"> & { type: "backlink" }) | undefined {
```
Comment (namjul) [inline packages/common-all/src/BacklinkUtils.ts:13]: Maybe extract an alias out of the return types. Something like `type BackDLink = Omit<DLink, "type"> & { type: "backlink" }`

File: packages/plugin-core/src/test/suite-integ/CopyNoteLink.test.ts (hunk 9 above)
Comment (tma66) [reply packages/plugin-core/src/test/suite-integ/CopyNoteLink.test.ts:369]: This test was failing on github every now and then; for some reason, the editor was still sometimes reading the previous document. I added this change in hopes that it fixes that.

File: packages/common-all/src/BacklinkUtils.ts (hunk 8 above)
Comment (tma66) [reply packages/common-all/src/BacklinkUtils.ts:37]: Definitely could happen. Do you think its worth the extra runtime to perform a dup check?

File: packages/unified/src/remark/utils.ts (hunk 7 above)
Comment (tma66) [reply packages/unified/src/remark/utils.ts:369]: Makes sense. I moved `BacklinksUtil` into common-all for this reason. Let's do it in another step of refactoring

File: packages/common-all/src/engine/EngineV3Base.ts (hunk 4 above)
Comment (tma66) [reply packages/common-all/src/engine/EngineV3Base.ts:454]: ah good reminder

File: packages/common-all/src/engine/EngineV3Base.ts (hunk 5 above)
Comment (tma66) [reply packages/common-all/src/engine/EngineV3Base.ts:461]: There are places where the note body is needed. Right now it'll be messier to change so leaving as is for now...

File: packages/common-all/src/engine/EngineV3Base.ts (hunk 1 above)
Comment (tma66) [reply packages/common-all/src/engine/EngineV3Base.ts:50]: Right now its part of `DCommonProps` so changing to `protected` isn't possible w/o another refactoring :/

File: packages/common-all/src/BacklinkUtils.ts (hunk 10 above)
Comment (tma66) [reply packages/common-all/src/BacklinkUtils.ts:13]: Sure, can do

**Comment by tma66** [comment]
> Wondering if this will maybe fix this issue? It's our most recent CROP selection: #1207
> 
> If not, then any changes I do on it I'll build off this review.

Actually this will fix that bug after this latest version. Just tested it

File: packages/common-all/src/BacklinkUtils.ts (hunk 8 above)
Comment (jonathanyeung) [reply packages/common-all/src/BacklinkUtils.ts:36]: I think it's worth doing the check (shouldn't be too expensive).  Otherwise we may have funky UI in the backlinks panel and potentially in other places as well.

File: packages/common-all/src/engine/EngineV3Base.ts (hunk 3 above)
Comment (kevinslin) [reply packages/common-all/src/engine/EngineV3Base.ts:447]: no, we should continue to remove it

we use `DNoteLink` in new methods, the key difference is the location of the `alias` field (see references of dlink2DNoteLink)
    - DLink has an `alias` field, it's not supposed to be on the link field but actually in the `from/to` field
    - `DNoteLoc` is a super set of `DLoc` (has an `alias` field)
    - `DNoteLink` is used in other places

File: packages/common-all/src/BacklinkUtils.ts [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+    }
+    return;
+  }
+
+  /** Adds a backlink by mutating the 'note' argument in place.
+   * Check if backlink already exists before pushing
+   *
+   *  @param note note that the link is pointing to. (mutated)
+   *  @param link backlink to add. */
+  static addBacklink({
```
Comment (kevinslin) [inline packages/common-all/src/BacklinkUtils.ts:31]: can we rename this to `addBacklinkInplace` to make it obvious?

**Comment by kevinslin** [comment]
looks like theres a compiler issue

**Review by kevinslin — APPROVED** [review]
